imqsauth.go

```diff
@@ -234,6 +251,221 @@ func createDB(config *authaus.Config) bool {
 	return true
 }
 
+func healthCheck(ic *auth.ImqsCentral, cleanAll bool) bool {
+	db := ic.Central.GetRoleGroupDB()
+	sb := strings.Builder{}
+	w := tabwriter.NewWriter(&sb, 0, 8, 1, '\t', tabwriter.AlignRight)
+	errorSet := map[string]error{}
+	_, err := fmt.Fprintf(w, "\n")
+	if err != nil {
+		// do nothing
+	}
+	fmt.Fprintf(w, "Health Check\n")
+	fmt.Fprintf(w, "------------\n")
+
+	// General info
+	// ------------
+
+	// Migration version
+	v, e := authaus.GetMigrationVersion(&ic.Config.Authaus.DB)
+	if e != nil {
+		fmt.Fprintf(w, "Could not read migration version...\n%v\n", e)
+	} else {
+		fmt.Fprintf(w, "Migration version:\t%v\n", v)
+	}
+
+	// Groups
+	groups, err := db.GetGroups()
+	if err == nil {
+		fmt.Fprintf(w, "Total nr of groups:\t%v\n", len(groups))
+	} else {
+		fmt.Fprintf(w, "Total nr of groups:\t%v\n", err)
+	}
+
+	// Users
+	users, err := ic.Central.GetAuthenticatorIdentities(authaus.GetIdentitiesFlagDeleted)
+	if err == nil {
+		fmt.Fprintf(w, "Total nr of users:\t%v\n", len(users))
+	} else {
+		fmt.Fprintf(w, "Total nr of users:\t%v\n", err)
+	}
+
+	// Non-archived users without sessions
+	tokens, e := ic.Central.GetAllTokens(false)
+	if e != nil {
+		fmt.Fprintf(w, "Could not retrieve tokens from session db\n")
+	} else {
+		fmt.Fprintf(w, "Active auth session tokens retrieved:\t%v\n", len(tokens))
+	}
+
+	oAuthTokenIDs, e := ic.Central.GetAllOAuthTokenIDs()
+	if e != nil {
+		fmt.Fprintf(w, "Could not retrieve tokens from session db\n")
+	} else {
+		fmt.Fprintf(w, "OAuth session tokens retrieved:\t%v\n", len(oAuthTokenIDs))
+	}
+
+	// detect any missing oauth sessions
+	nrSessionsWithOrphanedOAuthIds := findMissingOAuthTokens(oAuthTokenIDs, tokens)
+	fmt.Fprintf(w, "Nr of sessions with orphaned OAuth IDs:\t%v\n", nrSessionsWithOrphanedOAuthIds)
+
+	stats := map[string]int{}
+	stats[enabled] = 0
+	stats[admin] = 0
+	stats[archived] = 0
+	stats[unArchivedWithoutPermit] = 0
+
+	typeCounts := map[authaus.AuthUserType]*enabledTotal{
+		authaus.UserTypeDefault: {0, 0},
+		authaus.UserTypeLDAP:    {0, 0},
+		authaus.UserTypeOAuth:   {0, 0},
+		authaus.UserTypeMSAAD:   {0, 0},
+	}
+	missingGroups, errPerms2, _ := getUserStats(ic, users, stats, typeCounts, errorSet)
+	errFlag := ""
+	if errPerms2 {
+		errFlag = "*"
+	}
+	fmt.Fprintf(w, "Nr of enabled users:\t%v%v\n", stats[enabled], errFlag)
+	fmt.Fprintf(w, "Nr of admin users:\t%v%v\n", stats[admin], errFlag)
+	fmt.Fprintf(w, "Nr of archived users:\t%v\n", stats[archived])
```

Comment:
So if `ic.Central.GetAuthenticatorIdentities(authaus.GetIdentitiesFlagDeleted)` returns an error, then none of the code will break, but this printout will display `0` for everything, it feels like we should rather just not print anything if the variables we are depended on failed to populate.